### PR TITLE
Local file uploader

### PR DESCRIPTION
**Pull request by muskangarg98**
Local File Uploader Add-on to enable the user to provide a local file. 

To run this Local File uploader do turn on the Resources -> Advanced Google Services -> Drive API.

Also, i have added the convert file feature (which paneendra suggested) which now uploads the file according to it's mime-type.

### CODE REVIEW

File: form.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * Creates a menu entry in the Google Docs UI when the document is opened.
+ * This method is only used by the regular add-on, and is never called by
+ * the mobile add-on version.
+ *
+ * @param {object} e The event parameter for a simple onOpen trigger. To
+ *     determine which authorization mode (ScriptApp.AuthMode) the trigger is
+ *     running in, inspect e.authMode.
+ */
+function onOpen() {
+  DocumentApp.getUi()
```
Comment (paneendra-google) [inline form.js]: SpreadsheetApp since we will be attaching this with spreadsheet and not a doc, also in other places

File: form.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * This method is only used by the regular add-on, and is never called by
+ * the mobile add-on version.
+ *
+ * @param {object} e The event parameter for a simple onOpen trigger. To
+ *     determine which authorization mode (ScriptApp.AuthMode) the trigger is
+ *     running in, inspect e.authMode.
+ */
+function onOpen() {
+  DocumentApp.getUi() 
+      .createMenu('Form')
```
Comment (paneendra-google) [inline form.js]: This should be 'File Mapping' ?

File: form.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      .addItem('Open', 'Display')
```
Comment (paneendra-google) [inline form.js]: Open -> Edit

File: form.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      .addToUi();
+}
+
+
+/**
+ * Opens a form in the document containing the add-on's user interface.
+ */
+function Display() {
```
Comment (paneendra-google) [inline form.js]: DisplayFileMapping

File: form.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  var html = HtmlService.createHtmlOutputFromFile('uploadform')
```
Comment (paneendra-google) [inline form.js]: fileuploadform

File: form.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      .setWidth(600)
+      .setHeight(425);
+  DocumentApp.getUi() 
+      .showModalDialog(html, 'Upload a File');
```
Comment (paneendra-google) [inline form.js]: Upload a file to Google Drive

File: form.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ * from base64, then create a blob from all of these and then create a 
+ * file in the drive
+ *
+ * @param {string} data The base64 encoded file data.
+ * @param {string} mimetype The required mimetype of the file.
+ * @param {string} filename The required filename for the file. 
+ * @return {string} string containing the drive url of the file created.
+ */
+function processForm(data, mimetype, filename) {
+  var filedata = Utilities.base64Decode(data);
```
Comment (paneendra-google) [inline form.js]: are we expecting all data to be base64 encoded?

File: form.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  var blob = Utilities.newBlob(filedata);
+  
+  var driveFile = {
+    title: filename,
+    mimeType: mimetype
+  };
+  driveFile = Drive.Files.insert(driveFile, blob, {
+    convert: true
```
Comment (paneendra-google) [inline form.js]: Not sure if this should always be true. Should be true for Excel files only

File: form.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  });
+  
+  driveFile = DriveApp.getFolderById(driveFile.id);
+  
+  //Logger.log(mimetype);
```
Comment (paneendra-google) [inline form.js]: remove

File: uploadform.html [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      // Executed when the submit button is clicked
+      function handleFormSubmit(formObject) {
+        var file = formObject.myFile.files[0];
+        var fr = new FileReader();
+        fr.fileName = file.name;
+        fr.type = file.type;
+        
+        // Event handler executed when the file load event is fired
+        fr.onload = function(e) {
+
```
Comment (paneendra-google) [inline uploadform.html]: no need for empty line at start / end of a block

File: uploadform.html [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        var file = formObject.myFile.files[0];
+        var fr = new FileReader();
+        fr.fileName = file.name;
+        fr.type = file.type;
+        
+        // Event handler executed when the file load event is fired
+        fr.onload = function(e) {
+            
+            var data = e.target.result.replace(/^.*,/, '');
+            //var mimetype = e.target.result.match(/^.*(?=;)/)[0];
```
Comment (paneendra-google) [inline uploadform.html]: remove or uncomment

**Review by paneendra-google — CHANGES_REQUESTED** [review]

File: form.js (hunk 2 above)
Comment (muskangarg98) [reply form.js]: This is mainly for testing that's why didn't change it.

File: form.js (hunk 3 above)
Comment (muskangarg98) [reply form.js]: So, Display is the function to be executed when open is clicked.

File: form.js (hunk 8 above)
Comment (muskangarg98) [reply form.js]: and what about doc files?

File: form.js (hunk 7 above)
Comment (muskangarg98) [reply form.js]: Yes, the file type input in an html form return the file data in base64 decoded format.

File: form.js (hunk 8 above)
Comment (paneendra-google) [reply form.js]: We only support excel as of now

File: form.js (hunk 2 above)
Comment (paneendra-google) [reply form.js]: We could still keep useful names for testing as well :)

**Review by paneendra-google — CHANGES_REQUESTED** [review]

File: form.js (hunk 2 above)
Comment (muskangarg98) [reply form.js]: Yeah true !!

**Review by paneendra-google — COMMENTED** [review]
Let's move these into a separate folder?

**Review by paneendra-google — APPROVED** [review]

**Pull request merged**